File: StartUpExample/StartUpExampleJava/app/src/main/java/com/tinlone/startupexamplejava/dependencies/database/DatabaseProxy.java
package com.tinlone.startupexamplejava.dependencies.database;

import com.tinlone.startupexamplejava.dependencies.Logger;

/**
 * 假设使用代理模式做 插拔式DatabaseHelper
 */
public class DatabaseProxy {

    private static IDatabaseHelper dbHelper = null;

    private DatabaseProxy() {
        Logger.log("loglog", "DatabaseProxy initialized");
    }

    public static void initialize(IDatabaseHelper config) {
        // 仅做模拟
        dbHelper = config;
    }

    public static DatabaseProxy getInstance() {
        return Holder.INSTANCE;
    }

    public boolean insert(Object obj) {
        if (dbHelper == null) {
            throw new IllegalStateException("IDatabaseHelper has not initialize");
        }
        return dbHelper.insert(obj);
    }

    private static final class Holder {
        public static final DatabaseProxy INSTANCE = new DatabaseProxy();
    }

}

File: StartUpExample/StartUpExampleKotlin/app/src/main/java/com/tinlone/startupexamplekotlin/dependencies/database/DatabaseProxy.kt
package com.tinlone.startupexamplekotlin.dependencies.database

import com.tinlone.startupexamplekotlin.dependencies.Logger

/**
 * 假设使用代理模式做 插拔式DatabaseHelper
 */
class DatabaseProxy private constructor() {
    fun insert(obj: Any?): Boolean {
        checkNotNull(dbHelper) { "IDatabaseHelper has not initialize" }
        return dbHelper!!.insert(obj)
    }

    private object Holder {
        val instance = DatabaseProxy()
    }

    companion object {
        private var dbHelper: IDatabaseHelper? = null
        fun initialize(config: IDatabaseHelper?) {
            // 仅做模拟
            dbHelper = config
        }

        fun getInstance(): DatabaseProxy {
            return Holder.instance
        }
    }

    init {
        Logger.log("loglog", "DatabaseProxy initialized")
    }
}